File: frameworks/fw-playwright/tests/checkboxes.spec.js
// @ts-check
const { test, expect } = require('@playwright/test');

// Checkbox page with mixed locator strategies per test.
test.describe('Checkboxes @smoke', () => {
  test('checkbox toggles state (css + nth)', async ({ page }) => {
    await page.goto('/checkboxes');
    const boxes = page.locator('#checkboxes input');
    const first = boxes.nth(0);
    const second = boxes.nth(1);
    await expect(first).not.toBeChecked();
    await first.check();
    await expect(first).toBeChecked();
    await second.uncheck();
    await expect(second).not.toBeChecked();
  });

  test('both checkboxes found via getByRole', async ({ page }) => {
    await page.goto('/checkboxes');
    const boxes = page.getByRole('checkbox');
    await expect(boxes).toHaveCount(2);
    await boxes.first().check();
    await expect(boxes.first()).toBeChecked();
  });

  test('checkbox toggle via xpath locator', async ({ page }) => {
    await page.goto('/checkboxes');
    const second = page.locator('xpath=//form[@id="checkboxes"]/input[2]');
    await expect(second).toBeChecked(); // second box is checked by default
    await second.uncheck();
    await expect(second).not.toBeChecked();
  });
});

File: frameworks/fw-ruby/spec/checkboxes_spec.rb
require "spec_helper"

RSpec.describe "the-internet checkboxes", type: :feature do
  it "toggles checkbox 1 via css and checkbox 2 via xpath" do
    visit "/checkboxes"
    box1 = find(:css, "#checkboxes input:nth-of-type(1)")
    box2 = find(:xpath, "//form[@id='checkboxes']/input[2]")
    expect(box1).not_to be_checked
    expect(box2).to be_checked
    box1.click
    box2.click
    expect(find(:css, "#checkboxes input:nth-of-type(1)")).to be_checked
    expect(find(:xpath, "//form[@id='checkboxes']/input[2]")).not_to be_checked
  end

  it "polls until checkbox state settles after a toggle (polling loop)" do
    visit "/checkboxes"
    find(:css, "#checkboxes input:nth-of-type(1)").click
    poll_until(timeout: 20, interval: 1) do
      page.driver.browser
          .find_element(css: "#checkboxes input:nth-of-type(1)")
          .selected?
    end
    selenium_wait(timeout: 10) do |d|
      d.find_element(css: "#checkboxes input:nth-of-type(1)").selected?
    end
  end

  it "waits explicitly for the page heading via Selenium::WebDriver::Wait" do
    visit "/checkboxes"
    heading = selenium_wait(timeout: 20) do |d|
      el = d.find_element(xpath: "//div[@id='content']//h3")
      el if el.text == "Checkboxes"
    end
    expect(heading.text).to eq("Checkboxes")
  end
end
